rfd/0205-improved-onprem-joining.md

```diff
@@ -0,0 +1,1010 @@
+    rotate_on_next_renewal: false
+
+status:
+  bound_keypair:
+    # If `spec.onboarding.initial_public_key` is unset, this value will be
+    # generated server-side and made available here. If
+    # `spec.onboarding.initial_join_secret` is set, its value will be copied
+    # here.
+    initial_join_secret: <random>
+
+    # The public key of the bot associated with this token, set on first join.
+    bound_public_key: <data>
+
+    # The current bot instance UUID. A new UUID is issued on rejoin; the previous
+    # UUID will be linked via a `previous_instance_id` in the bot instance.
+    bound_bot_instance_id: <uuid>
+
+    # A count of remaining rejoins; if `.spec.bound_keypair.rejoining.total_rejoins`
+    # is incremented, this value will be incremented by the same amount. If
+    # decremented, this value cannot fall below zero. If
+    # `.spec.bound_keypair.rejoining.unlimited` is set, this value will always
+    # be 0 but rejoin attempts will succeed.
+    remaining_rejoins: 10
+
+    # A count of the total number of joins performed using this token.
+    join_count: 0
+
+    # The timestamp of the last successful joining or rejoining attempt, if any.
+    last_joined_at: null
+
+    # The timestamp of the last successful keypair rotation, if any.
+    last_rotated_at: null
+```
+
+#### Proto Draft
+
+```protobuf
+message RegisterUsingBoundKeypairInitialRequest {
+  types.RegisterUsingTokenRequest join_request = 1;
+
+  // If set, requests a rotation to the new public key. The joining challenge
+  // must first be completed using the previous key, and upon completion a new
+  // challenge will be issued for this key. Certificates will only be returned
+  // after the second challenge is complete.
+  bytes new_public_key = 2;
+
+  // A document signed by Auth containing join state parameters from the
+  // previous join attempt. Not required on initial join; required on all
+  // subsequent joins.
+  bytes previous_join_state = 3;
+}
+
+message RegisterUsingBoundKeypairChallengeResponse {
+  bytes solution = 1;
+}
+
+message RegisterUsingBoundKeypairRequest {
+  oneof payload {
+    RegisterUsingBoundKeypairInitialRequest init = 1;
+    RegisterUsingBoundKeypairChallengeResponse challenge_response = 2;
+  }
+}
+
+message RegisterUsingBoundKeypairCertificates {
+  // Signed Teleport certificates resulting from the join process.
+  Certs certs = 1;
+
+  // A signed join state document to be provided on the next join attempt.
+  bytes join_state = 2;
+}
+
+message RegisterUsingBoundKeypairResponse {
+  oneof response {
+    // A challenge to sign. During keypair rotation, a second challenge will be
+    // provided to verify the new keypair before certs are returned.
+    string challenge = 1;
+    RegisterUsingBoundKeypairCertificates certs = 2;
+  }
+}
+
+service JoinService {
+  // ...snip...
+
+  rpc RegisterUsingBoundKeypair(stream RegisterUsingBoundKeypairRequest) returns (stream RegisterUsingBoundKeypairResponse);
```

Comment:
nit: What's the benefit of this being a streaming RPC rather than a series of unary RPCs?

Reply:
Just to write down some reasoning from the verbal discussion:
- Most other methods with a challenge-response process (e.g. TPM joining) do this already
- Consistently ties a renewal ceremony to one Auth instance if behind a LB, which might be important for consistency on some backends
- Is mildly nicer to keep things within a single socket session when multiple (potentially optional) steps may be involved in a join attempt, e.g. keypair rotation.